### PR TITLE
Remove DatasetProvider from ModelEvaluator api's.

### PR DESCRIPTION
**Pull request by lovettchris**
# Pull Request Template

## Description

In the case of a distributed partial training evaluator based on AML backend we don't want the full cost of downloading and preparing a dataset on every "send". So removing that from the api makes it more clear that the dataset will not change randomly during every send call on an `AsyncModelEvaluator` or during each evaluate call on a `ModelEvaluator`.

[9:19 AM] Piero Kauffmann
    
I like this idea. One other advantage I see from that approach is that it also removes the need to pass DatasetProvider object to the search algorithms, which is a bit annoying right now ([Search Algorithms — Archai (microsoft.github.io)](https://microsoft.github.io/archai/reference/api/archai.discrete_search.algos.html)). This also allows the user to choose a custom dataset provider for each evaluator if they want (i.e. use low-res images for evaluator1 and high-res images for evaluator2, or something like that). Do you want to start a PR? I can also help if you want to.

P.S If we change this interface this won't matter at all, but I'm a bit curious why the MNIST dataset provider download issue is happening at all, because Mnist torchvision dataset class (used in MnistDatasetProvider in your example) will not download the dataset during obj creation if it finds the dataset in the specified root directory, so it should not re-download anything even if you create a new dataset provider object multiple times (if you pass the same root directory).

## Changes

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (causes existing functionality to not work as expected)
- [ ] Documentation update

**Configuration**:
* Archai version:
* Environment:
* OS:

## Final checks:

- [x] Follows guidelines of project
- [x] Self-review of code
- [x] Commented code in hard-to-understand areas
- [x] Corresponding changes to the documentation
- [x] Changes generate no new warnings
- [x] Dependent changes have already been merged
- [x] Checked code correction any misspellings

### CODE REVIEW

**Review by piero2c — APPROVED** [review]
LGTM!

File: archai/common/common.py
Comment (piero2c) [inline archai/common/common.py:39]: Can you review these changes, @gugarosa?

Comment (gugarosa) [reply archai/common/common.py:39]: Sure, looks good! This was a typo from a previous PR.

`get_inst()` is the correct method.

Comment (lovettchris) [reply archai/common/common.py:39]: thanks for confirming.

File: archai/discrete_search/api/search_objectives.py
Comment (lovettchris) [reply archai/discrete_search/api/search_objectives.py:57]: Done

**Comment by gugarosa** [comment]
Please feel free to merge it @lovettchris. If any of the tests fails, I will correct them after the merge.

**Comment by lovettchris** [comment]
@piero2c, I fixed the AmlTrainingValAccuracy send method, and will merge this when the tests above finish green.

**Pull request merged**
